simplified-app-shared/src/main/java/org/nypl/simplified/app/ReaderActivitySyncManager.kt

```diff
@@ -0,0 +1,173 @@
+package org.nypl.simplified.app
+
+import android.app.AlertDialog
+import android.content.Context
+import com.io7m.jfunctional.Some
+import org.nypl.simplified.app.reader.ReaderBookLocation
+import org.nypl.simplified.app.utilities.UIThread
+import org.nypl.simplified.books.core.AccountCredentials
+import org.nypl.simplified.books.core.AccountsControllerType
+import org.nypl.simplified.multilibrary.Account
+import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry
+import org.readium.sdk.android.Package
+import org.slf4j.LoggerFactory
+import java.net.URI
+import java.util.Timer
+import kotlin.concurrent.schedule
+
+
+interface ReaderSyncManagerDelegate {
+  fun navigateToLocation(location: ReaderBookLocation)
+  fun bookmarkUploadDidFinish(bookmark: NYPLBookmark, bookID: String)
+  fun bookmarkSyncDidFinish(success: Boolean, bookmarks: List<NYPLBookmark>)
+}
+
+/**
+ * Provide support for sync-related activities and logic.
+ * Utilize ReaderSyncManagerDelegate Interface to
+ * communicate commands to a book renderer.
+ */
+class ReaderSyncManager(private val entryID: String,
+                        credentials: AccountCredentials,
+                        private val libraryAccount: Account,
+                        private val delegate: ReaderSyncManagerDelegate) : ReaderSyncManagerDelegate by delegate
+{
+  private companion object {
+    val LOG = LoggerFactory.getLogger(ReaderSyncManager::class.java)!!
+  }
+
+  //TODO Work In Progress -
+  //TODO observe when set, then show jump to location dialog with appropriate messaging
+  val bookPackage: Package? = null
+
+  // Delay server posts until first page sync is complete
+  private var delayPageSync: Boolean = true
+  private val annotationsManager = AnnotationsManager(libraryAccount, credentials, delegate as Context)
+  private var queueTimer = Timer()
+
+  /**
+   * See if sync is enabled on the server before attempting to synchronize
+   * bookmarks or reading position. Set sharedPrefs accordingly.
+   * Execute the closure if it's already on or successfully enabled on the server.
+   */
+  fun serverSyncPermission(account: AccountsControllerType,
+                           completion: () -> Unit)
```

Comment:
if two callers both hit this method, will we fire two requests? or will we store completion blocks and just call them when the outstanding request is completed?

Reply:
See below